src/keyboard/KeyboardViewport.tsx

```diff
@@ -0,0 +1,153 @@
+  className?: string;
+}>;
+
+const KEYMAP_SCALE = "keymap:scale";
+const DEFAULT_SCALE = window.localStorage.getItem(KEYMAP_SCALE) ?? "1";
+
+export const KeyboardViewport: FC<KeyboardViewportType> = ({
+  children,
+  className,
+}) => {
+  const targetRef = useRef<HTMLDivElement>(null);
+  const scaleRef = useRef<HTMLInputElement>(null);
+
+  const setScale = (param: "increase" | "decrease") => {
+    if (!targetRef.current || !scaleRef.current) return;
+
+    const current = scaleRef.current.value;
+
+    if (param === "increase" && Number(current) < 2) {
+      scaleRef.current.value = String(Number(scaleRef.current.value) + 0.2);
+    }
+
+    if (param === "decrease" && Number(current) > 0.2) {
+      scaleRef.current.value = String(Number(scaleRef.current.value) - 0.2);
+    }
+
+    localStorage.setItem(KEYMAP_SCALE, scaleRef.current.value);
+    targetRef.current.style.setProperty(
+      "transform",
+      `scale(${scaleRef.current.value})`,
+    );
+  };
+
+  const resetScale = () => {
+    if (!targetRef.current || !scaleRef.current) return;
+    targetRef.current.style.translate = "unset";
+    targetRef.current.style.setProperty("transform", "scale(1)");
+    scaleRef.current.value = "1";
+    localStorage.setItem(KEYMAP_SCALE, "1");
+  };
+
+  useEffect(() => {
+    if (!targetRef.current) return;
+
+    const target = targetRef.current;
+    const offset = { x: 0, y: 0 };
+    let isPanningActive = false;
+
+    function panStart(e: KeyboardEvent) {
+      if (e.key !== " ") return;
+      e.preventDefault();
+
+      target.style.cursor = "grab";
+      isPanningActive = true;
+    }
+
+    function panEnd(e: KeyboardEvent) {
+      if (e.key !== " ") return;
+      isPanningActive = false;
+      target.style.cursor = "unset";
+    }
+
+    function panMove(e: PointerEvent) {
+      if (!isPanningActive) return;
+      offset.x += e.movementX;
+      offset.y += e.movementY;
+      target.style.translate = `${offset.x}px ${offset.y}px`;
+    }
+
+    document.addEventListener("keydown", panStart);
+    document.addEventListener("keyup", panEnd);
+    target.addEventListener("pointermove", panMove);
+
+    return () => {
+      document.removeEventListener("keydown", panStart);
+      document.removeEventListener("keyup", panEnd);
+      target.removeEventListener("pointermove", panMove);
+    };
+  }, []);
+
+  useEffect(() => {
+    if (!scaleRef.current || !targetRef.current) return;
+
+    const input = scaleRef.current;
+    const target = targetRef.current;
+
+    input.value = DEFAULT_SCALE;
+    target.style.setProperty("transform", `scale(${DEFAULT_SCALE})`);
+
+    function onInputChange(e: Event) {
+      const value = (e.currentTarget as HTMLInputElement).value;
+      target.style.setProperty("transform", `scale(${value})`);
+      localStorage.setItem(KEYMAP_SCALE, value);
+    }
+
+    input.addEventListener("change", onInputChange);
+    return () => {
+      input.removeEventListener("change", onInputChange);
```

Comment:
I think I noticed from your video that it only seems to change once you've let go. IIRC if you use react state + a on change handler in JSX it will update as you drag (which seems desirable). Overall not a fan of doing all this imperative work, which IMO should only be done when required (refs, event listeners, etc)

Reply:
revised the implementation